Substantial rewrite of the Cookbook guide (#435)
ProjectPythia/projectpythia.github.io: portal/cookbook-guide.md

However, we recommend transfering to the Pythia organization early, for a few reasons:
- Fewer settings to tweak later
- Easier to get help from the Pythia infrastructure team
- Encourages collaboration

```{Note}
You're still in control! You will always retain write access to your Cookbook repository even after transfering ownership to the Pythia organization.

Also, _don't worry about breaking anything!_ Your repo will not affect any other Project Pythia content until you initiate the request to list it on the [Cookbook Gallery](https://cookbooks.projectpythia.org) (see below...)
```

### Steps to transfer the repository

1. [Contact the Pythia team via the Pangeo Discourse](https://discourse.pangeo.io/c/education/project-pythia/60) (or otherwise) to let us know about your cookbook.
1. You will get an invitation to join the the [ProjectPythia organization](https://github.com/ProjectPythia).
1. Accept the GitHub invitation. _Welcome to the team!_
1. Once you are a member of the organization, navigate to the Settings of your cookbook repository and scroll down to the **Danger Zone**.
1. Click "Transfer".
1. Select or type "ProjectPythia", confirm, and transfer.

> Once you have successfully transfered the repository, you'll most likely want to make a [personal fork and a local clone of the repository](https://foundations.projectpythia.org/foundations/github/github-cloning-forking.html) so that you can continue to develop and collaborate on the Cookbook via the [forking workflow](https://foundations.projectpythia.org/foundations/github/github-workflows.html#forking-workflow).

Reviewer: After creating a personal fork of the repo I subsequently learned that GitHub had disabled all of the workflows and I had to re-enable them.

Author: > After creating a personal fork of the repo I subsequently learned that GitHub had disabled all of the workflows and I had to re-enable them.

This is expected! Normally we only want the workflows to operate on the upstream fork hosted in the ProjectPythia org. The typical and supported way to preview the changes that you make in your personal fork is to open a pull request on the upstream fork.

I can add this as a clarification in the guide.

There *may* be use cases where it makes sense to enable the workflows in the personal fork, but honestly I have never needed this for own cookbook development work.

Reviewer: Gotcha. Clarifying would help.

Author: Clarification is now in the note block at the end of [this section](https://projectpythia.org/_preview/435/cookbook-guide.html#where-is-my-book).

### Customizing your environment file

1. Within `environment.yml` (in the root of the repository), change `name` from `cookbook-dev` to `<your-cookbook-name>-dev` (e.g. `cesm-cookbook-dev`) and add all required libraries and other dependencies under `dependencies:`. Commit the changes.
1. Create the [conda environment](https://foundations.projectpythia.org/foundations/conda.html) with `conda env create -f environment.yml`. If it crashes, try running `conda config --set channel_priority strict`
1. Activate your environment with `conda activate <env-name>`

You're now ready to create and run awesome notebooks.

### Customizing your GitHub actions

Your repository includes automation for building and publishing your Cookbook, powered by [GitHub Actions](https://docs.github.com/en/actions). Now that you have created a custom name for your conda environment (`<your-cookbook-name>-dev`), you need to edit three files found in the `.github/workflows` section of your repo:
- `.github/workflows/nightly-build.yaml`
- `.github/workflows/publish-book.yaml`
- `.github/workflows/trigger-book-build.yaml`

> In each of these files, in the field called `environment_name:`, replace  `cookbook-dev` with the name your used in your `environment.yml` file (probably `<your-cookbook-name>-dev`). Commit these changes.

Reviewer: ```suggestion
In each of these files, in the field called `environment_name:`, replace  `cookbook-dev` with the name you used in your `environment.yml` file (probably `<your-cookbook-name>-dev`). Commit these changes.
```

```{Note}
If these workflow files look mysterious and you don't know anything about how GitHub Actions work, don't worry! The Pythia team will help with any problems that arise with the Cookbook automation.
```

## Develop your cookbook

### Add notebooks

To add content, you should edit (and duplicate as necessary) the notebook template `notebooks/notebook-template.ipynb`. Using this template will help keep your content consistent with the expected Cookbook style.

You can add subfolders to organize your notebook files as necessary. However, the organization of the notebooks within the published book is dictated by the table of contents file `_toc.yml`. _Edit this file to add your chapters._ Take a look at the [`radar-cookbook/_toc.yml`](https://github.com/ProjectPythia/radar-cookbook/blob/main/_toc.yml) for example syntax, or consult the [JupyterBook documentation](https://jupyterbook.org/en/stable/structure/toc.html).

### Customize your home page

The file `README.md` serves as the homepage of your Cookbook, so it should be descriptive. Edit this file as appropriate, following guidance in the file itself.

### Build your Cookbook locally

> You should be able to build your new Cookbook on your local machine by doing

Reviewer: ```suggestion
You should be able to build your new Cookbook on your local machine by running this command from the root of the repository:
```

```
open _build/html/index.html
```

### Strip output from your notebooks before committing

Wherever feasible, we recommend only committing bare, unexecuted notebooks into your Cookbook repository. There are a few reasons:
- The notebooks are executed automatically each time the book is built
- Pre-existing output in the notebooks will have no effect on the published book
- Better reproducibility of your content
- The repo stays cleaner and leaner. Diffs of notebook changes are easier to interpret.

To strip a notebook, in a Jupyter Lab session, go to "Kernel" &rarr; "Restart Kernel and Clear Output of All Cells...". Save your notebook, then commit.

You're ready to push content up to GitHub and trigger the automated publishing pipeline.

## Deploying your Cookbook

Pythia Cookbooks are not just collections of notebooks! They are backed by GitHub-based automation that handles building, testing, previewing, and publishing the books online.

### About the publishing pipeline

Any time you push new content to the `main` branch of your Cookbook repo, actions are triggered that do the following:

> - create an up-to-date build environment using your `environment.yml` file
> - execute all the notebooks in that environment
> - package the book into a styled website
> - check all the links in the website to make sure they are valid
> - deploy the styled website so it's visible to the world

Reviewer 2: ```suggestion
- Create an up-to-date build environment using your `environment.yml` file
- Execute all the notebooks in that environment
- Package the book into a styled website
- Check all the links in the website to make sure they are valid
- Deploy the styled website so it's visible to the world
```

If the build is successful, you should be able to preview the result in a web browser by running
```
open _build/html/index.html
```

### Strip output from your notebooks before committing

Wherever feasible, we recommend only committing bare, unexecuted notebooks into your Cookbook repository. There are a few reasons:
- The notebooks are executed automatically each time the book is built
- Pre-existing output in the notebooks will have no effect on the published book
- Better reproducibility of your content
- The repo stays cleaner and leaner. Diffs of notebook changes are easier to interpret.

To strip a notebook, in a Jupyter Lab session, go to "Kernel" &rarr; "Restart Kernel and Clear Output of All Cells...". Save your notebook, then commit.

You're ready to push content up to GitHub and trigger the automated publishing pipeline.

## Deploying your Cookbook

Pythia Cookbooks are not just collections of notebooks! They are backed by GitHub-based automation that handles building, testing, previewing, and publishing the books online.

### About the publishing pipeline

> Any time you push new content to the `main` branch of your Cookbook repo, actions are triggered that do the following:

Reviewer: So "your Cookbook repo" is a little misleading here? The actions are only triggered on the cookbook that is in the ProjectPythia organization, not your fork of the repo, correct?

Author: I added a note block at the bottom of [this section](https://projectpythia.org/_preview/435/cookbook-guide.html#where-is-my-book) to try to clarify this. 

We can use more precise language in the sentence you highlighted. I didn't want it to get too convoluted. Strictly speaking, it's the `main` branch of the `upstream` repository, not your personal fork. Unless you *haven't* transferred ownership of your repo, it which case it would be the `main` branch of your personal repository (which is not a fork and thus does not have an upstream equivalent).

Author: Any suggested language for this?

Reviewer: Nope. That looks good.

### About the publishing pipeline

Any time you push new content to the `main` branch of your Cookbook repo, actions are triggered that do the following:
- Create an up-to-date build environment using your `environment.yml` file
- Execute all the notebooks in that environment
- Package the book into a styled website
- Check all the links in the website to make sure they are valid
- Deploy the styled website so it's visible to the world

### Where is my book?

The URL of your published book will depend on where the source repository is housed.
- If the repo is in the [Project Pythia organization](https://github.com/ProjectPythia), the book is published to `https://projectpythia.org/[YOUR_COOKBOOK_REPO_NAME]/`
- If the repo is in your personal GitHub space, the book is published to `https://[YOUR_GITHUB_USERNAME].github.io/[YOUR_COOKBOOK_REPO_NAME]/`

Here's a handy trick for finding your published book:
- On the home page of your GitHub repo, look for the "About" section in the right-hand side bar.
- Click the "gear" icon next to "About"
- Select the checkbox "Use your GitHub Pages website".

> - Is that checkbox missing? That likely means you're looking at the "About" section of your _personal fork_, not the upstream fork. Navigate back to `https://github.com/ProjectPythia/[YOUR_COOKBOOK_REPO_NAME]/` and look for it there.

Author: @clyne I added one last clarification here

Author: I should have merged the PR while I had approval, lol. If you have a chance to approve again, that would be terrific.

Reviewer: Done!

1. Focus on representative subsets of data that can be packaged alongside the cookbook in-repo. An example is the [Landsat ML Cookbook](https://projectpythia.org/landsat-ml-cookbook/README.html)
1. Discuss your larger data storage needs with the Pythia team. We are currently experimenting with cloud object storage for Cookbooks via NSF JetStream2.
1. Provide the tools and/or clear documentation for accessing the data that you have stored somewhere else

---

## Use the template

1. If you don't already have a GitHub account, create one by following the [Getting Started with GitHub guide](https://foundations.projectpythia.org/foundations/getting-started-github.html)
1. On the [Cookbook Template repository](https://github.com/ProjectPythia/cookbook-template), click "Use this template &rarr; Create a new repository"
1. Give your repository a descriptive name followed by `-cookbook` (e.g., `hydrology-cookbook`, `hpc-cookbook`, `cesm-cookbook`) and a description

> 1. Choose "Include all branches" and create the repository. Your browser will be directed to the newly created repository under your GitHub account

Reviewer: Small nit, but the option to "include all branches" occurs above the repository name selection

Author: Resolved: https://projectpythia.org/_preview/435/cookbook-guide.html#use-the-template

> 1. Under Settings &rarr; Pages, ensure that GitHub Pages is enabled. If it is not, change the Branch from "None" to "gh-pages/(root)" and click "Save"

Reviewer: ```suggestion
1. Under Settings &rarr; Pages, ensure that GitHub Pages is enabled by checking that `Branch` is set to "gh-pages", and the folder set to "gh-pages/(root)". If it is not, change the Branch from "None" to "gh-pages/(root)" and click "Save"
```

> 1. Under Settings &rarr; Actions &rarr; General, allow Github Actions to **push** to the repository <img width="901" alt="Screenshot 2023-01-13 at 3 12 47 PM" src="https://user-images.githubusercontent.com/26660300/212428991-cd0ae2f0-73ca-40d8-b983-f122359463aa.png">

Reviewer: The word "push" does not appear the page. The included image suggest that "Read and write permissions", under Workflow permissions", should be selected. Is this an additional step, or what is meant by allowing "push"?

Author: I'll clarify the text. We mean that "Read and write permissions" should be selected. No additional steps.

Author: Resolved: https://projectpythia.org/_preview/435/cookbook-guide.html#use-the-template

However, we recommend transfering to the Pythia organization early, for a few reasons:
- Fewer settings to tweak later
- Easier to get help from the Pythia infrastructure team
- Encourages collaboration

```{Note}
You're still in control! You will always retain write access to your Cookbook repository even after transfering ownership to the Pythia organization.

Also, _don't worry about breaking anything!_ Your repo will not affect any other Project Pythia content until you initiate the request to list it on the [Cookbook Gallery](https://cookbooks.projectpythia.org) (see below...)
```

### Steps to transfer the repository

1. [Contact the Pythia team via the Pangeo Discourse](https://discourse.pangeo.io/c/education/project-pythia/60) (or otherwise) to let us know about your cookbook.
1. You will get an invitation to join the the [ProjectPythia organization](https://github.com/ProjectPythia).
1. Accept the GitHub invitation. _Welcome to the team!_
1. Once you are a member of the organization, navigate to the Settings of your cookbook repository and scroll down to the **Danger Zone**.
1. Click "Transfer".

> 1. Select or type "ProjectPythia", confirm, and transfer.

Reviewer: There is another step following this one that asks which teams to give access to. Options are nobody, core, education, infrastructure, outreach, and "find a team". What is the guidance here?

Author: Hmm I guess the guidance should be to select "core". I think that includes all the active members of both "education" and "infrastructure". We haven't been very careful or consistent about how we use those teams.

Author: I selected "core" when I transferred a [new cookbook repo](https://github.com/ProjectPythia/radiative-feedback-cookbook) the other day, which is what got me started on this guide revision.

I forgot to include this guidance in the text.

Author: Resolved: https://projectpythia.org/_preview/435/cookbook-guide.html#steps-to-transfer-the-repository
